clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -1354,6 +1354,183 @@ test_synthetic()
     }
 }
 
+static void
+test_synthetic_with_syscall_seq()
+{
+    std::cerr << "\n----------------\nTesting synthetic with syscall sequences\n";
+    static constexpr int NUM_INPUTS = 7;
+    static constexpr int NUM_OUTPUTS = 2;
+    static constexpr int NUM_INSTRS = 9;
+    static constexpr int QUANTUM_DURATION = 3;
+    // We do not want to block for very long.
+    static constexpr double BLOCK_SCALE = 0.01;
+    static constexpr uint64_t BLOCK_THRESHOLD = 100;
+    static constexpr memref_tid_t TID_BASE = 100;
+    static constexpr uint64_t KERNEL_CODE_OFFSET = 123456;
+    std::vector<trace_entry_t> inputs[NUM_INPUTS];
+    for (int i = 0; i < NUM_INPUTS; i++) {
+        memref_tid_t tid = TID_BASE + i;
+        inputs[i].push_back(make_thread(tid));
+        inputs[i].push_back(make_pid(1));
+        inputs[i].push_back(make_version(TRACE_ENTRY_VERSION));
+        inputs[i].push_back(make_timestamp(10)); // All the same time priority.
+        for (int j = 0; j < NUM_INSTRS; j++) {
+            inputs[i].push_back(make_instr(42 + j * 4));
+            // Test accumulation of usage across voluntary switches.
+            if ((i == 0 || i == 1) && j == 1) {
+                inputs[i].push_back(make_timestamp(20));
+                inputs[i].push_back(make_marker(TRACE_MARKER_TYPE_SYSCALL, 42));
+                inputs[i].push_back(
+                    make_marker(TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL, 0));
+                inputs[i].push_back(make_timestamp(120));
+            }
+            // Test a syscall sequence starting at each offset within a quantum
+            // of instrs.
+            if (i <= QUANTUM_DURATION && i == j) {
+                inputs[i].push_back(make_marker(TRACE_MARKER_TYPE_SYSCALL, 84));
+                inputs[i].push_back(
+                    make_marker(TRACE_MARKER_TYPE_SYSCALL_TRACE_START, 84));
+                for (int k = 0; k <= QUANTUM_DURATION; ++k)
+                    inputs[i].push_back(make_instr(KERNEL_CODE_OFFSET + k));
+                inputs[i].push_back(make_marker(TRACE_MARKER_TYPE_SYSCALL_TRACE_END, 84));
+            }
+        }
+        inputs[i].push_back(make_exit(tid));
+    }
+    // A has a syscall sequence at [2,5], B has it at [3,6], C has it at [4,7],
+    // D has it at [5,8].
+    // The syscall sequence consists of 4 instrs which is greater than the
+    // #instr quanta.
+    // Total instrs in A, B, C, and D are 9 + 4 == 13, others have just 9.
+
+    // Hardcoding here for the 2 outputs and 7 inputs.
+    // We make assumptions on the scheduler's initial runqueue assignment
+    // being round-robin, resulting in 4 on core0 (odd parity letters) and 3 on
+    // core1 (even parity letters).
+    // The dots are markers and thread exits.
+    // A has a voluntary switch at its 6th instr (1st in that scheduling). Its
```

Comment:
After the 6th instr is here:  `A..AAAA...GA....GA^^^^AC...`.  Do you mean before the 6th instr?

Reply:
```
if ((i == 0 || i == 1) && j == 1) {
```

inserts the blocking syscall at j=1.  However, for A (i==0), at j=0 we add 1 regular + 4 syscall instrs. So the blocking syscall ends up at the 6th instr ordinal.


"..A..AAAA...CCC..CCCC...EEE..GGGA....CCCEEEGGGAACCC.EEE.AAAAA."
------------------------------------^

Pointed out the 6th instr for A, after which there's a switch to C

Reply:
I think I miscounted: the 4 A's in a row blurred into 3 or something (probably b/c all the tests generally only have 3 in a row with a quantum of 3).

Reply:
Maybe independent: but the new TRACE_MARKER_TYPE_SYSCALL added doesn't have a preceding instr... need to fix that in the test case.

Reply:
Made the test check more conditions, and added more detailed narration to make it clearer.

Reply:
Made the fallback check a bit stronger for the #instr quantum sequence.